server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/RareTermsAggregatorFactory.java

```diff
@@ -0,0 +1,165 @@
+import org.elasticsearch.search.aggregations.AggregationExecutionException;
+import org.elasticsearch.search.aggregations.Aggregator;
+import org.elasticsearch.search.aggregations.AggregatorFactories;
+import org.elasticsearch.search.aggregations.AggregatorFactory;
+import org.elasticsearch.search.aggregations.InternalAggregation;
+import org.elasticsearch.search.aggregations.NonCollectingAggregator;
+import org.elasticsearch.search.aggregations.pipeline.PipelineAggregator;
+import org.elasticsearch.search.aggregations.support.ValuesSource;
+import org.elasticsearch.search.aggregations.support.ValuesSourceAggregatorFactory;
+import org.elasticsearch.search.aggregations.support.ValuesSourceConfig;
+import org.elasticsearch.search.internal.SearchContext;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+public class RareTermsAggregatorFactory extends ValuesSourceAggregatorFactory<ValuesSource, RareTermsAggregatorFactory> {
+    private final IncludeExclude includeExclude;
+    private final int maxDocCount;
+
+    RareTermsAggregatorFactory(String name, ValuesSourceConfig<ValuesSource> config,
+                                      IncludeExclude includeExclude,
+                                      SearchContext context,
+                                      AggregatorFactory<?> parent, AggregatorFactories.Builder subFactoriesBuilder,
+                                      Map<String, Object> metaData, int maxDocCount) throws IOException {
+        super(name, config, context, parent, subFactoriesBuilder, metaData);
+        this.includeExclude = includeExclude;
+        this.maxDocCount = maxDocCount;
+    }
+
+    @Override
+    protected Aggregator createUnmapped(Aggregator parent, List<PipelineAggregator> pipelineAggregators, Map<String, Object> metaData)
+        throws IOException {
+        final InternalAggregation aggregation = new UnmappedRareTerms(name, pipelineAggregators, metaData);
+        return new NonCollectingAggregator(name, context, parent, factories, pipelineAggregators, metaData) {
+            @Override
+            public InternalAggregation buildEmptyAggregation() {
+                return aggregation;
+            }
+        };
+    }
+
+    @Override
+    protected Aggregator doCreateInternal(ValuesSource valuesSource, Aggregator parent, boolean collectsFromSingleBucket,
+                                          List<PipelineAggregator> pipelineAggregators, Map<String, Object> metaData) throws IOException {
+        if (collectsFromSingleBucket == false) {
+            return asMultiBucketAggregator(this, context, parent);
+        }
+        if (valuesSource instanceof ValuesSource.Bytes) {
+            ExecutionMode execution = ExecutionMode.MAP; //TODO global ords not implemented yet, only supports "map"
+
+            DocValueFormat format = config.format();
+            if ((includeExclude != null) && (includeExclude.isRegexBased()) && format != DocValueFormat.RAW) {
```

Comment:
I think the `DocValueFormat.RAW` check is being used to determine that the field used is a string field. But I see a few issues here (unless I'm misunderstanding what this is doing):
* Users can apply custom formats to non-string fields
* The `valuesSource` has already been checked above to be a ValuesSource.Bytes so this can only be a string field here?